File: pset2/crack.c
/**
 * PSet 2.3 - Crack
 * Written by Sudipto Ghosh for CS50
 */

#include <cs50.h>
#include <stdio.h>
#include <crypt.h>
#include <string.h>

const int ALPHABETS = 52;
const char *ALPHABET = "\0ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

int main(int argc, char **argv)
{
    // check validity of argument
    if (argc != 2 || argv[1] == NULL)
    {
        // print usage information and exit
        printf("Usage: ./crack hash\n");
        return 1;
    }
    
    // extract salt from hashed password
    char salt[] = {argv[1][0], argv[1][1], '\0'};
    
    // initialize attempt counter
    int attemptCount = 0;
    
    // initialize password with null characters
    char password[6] = {'\0'};
    
    // iterate over all permutations of ALPHABET
    for (int i = 0; i < ALPHABETS; i++)
    {
        for (int j = 0; j < ALPHABETS; j++)        
        {
            for (int k = 0; k < ALPHABETS; k++)
            {
                for (int l = 0; l < ALPHABETS; l++)        
                {
                    for (int m = 1; m < ALPHABETS; m++)        
                    {
                        // construct password for each permutation
                        password[0] = ALPHABET[m]; // first letter
                        password[1] = ALPHABET[l]; // second letter
                        password[2] = ALPHABET[k]; // third letter
                        password[3] = ALPHABET[j]; // fourth letter
                        password[4] = ALPHABET[i]; // fifth letter
                        
                        // increment attempt counter
                        attemptCount++; 
                        
                        // print password if match found and exit
                        if (!strcmp(crypt(password, salt), argv[1]))
                        {
                            printf("Password: %s\n", password);
                            printf("Attempts: %i\n", attemptCount);
                            return 0;
                        }
                    }
                }
            } 
        }
    }
    
    // print error message if password not cracked
    printf("Password could not be cracked\n");
    
    return 2;
}

File: pset6/crack/crack.py
"""
PSet 6 - Sentimental/Crack
Written by Sudipto Ghosh for CS50

"""

from time import sleep
from crypt import crypt
from sys import argv

ALPHABETS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"


def getArgs():
    """
    validates arguments passed during execution
    """
    # ensures only two arguments are passed
    if not len(argv) == 2:
        exit("Usage: python crack.py hash")

    return argv[1]


def check(password, salt, hash):
    """
    checks whether the password is valid
    """
    return hash == crypt(password, salt)


def crack(hash):
    """
    bruteforces hash to find out the password
    """
    # extract salt
    salt = hash[:2]

    # iterate over all possible permutations
    password = ""
    attempts = 0
    for i in range(len(ALPHABETS)):
        for j in range(len(ALPHABETS)):
            for k in range(len(ALPHABETS)):
                for l in range(len(ALPHABETS)):
                    for m in range(len(ALPHABETS)):
                        attempts += 1
                        if check(password, salt, hash):
                            return password, attempts, True
                        password = ALPHABETS[m] + password[1:]
                    password = password[:1] + ALPHABETS[l] + password[2:]
                password = password[:2] + ALPHABETS[k] + password[3:]
            password = password[:3] + ALPHABETS[j] + password[4:]
        password = password[:4] + ALPHABETS[i] + password[5:]

    return "", attempts, False


def main():
    # get cli arguments
    hash = getArgs()

    # get password and cracked flag
    password, attempts, cracked = crack(hash)

    # print password if cracked flag is true
    print(f"Password: {password}\nAttempts: {attempts}" if cracked else "Couldn't crack")


if __name__ == "__main__":
    main()